runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -394,6 +370,163 @@ private StreamingDataflowWorker(
+        new StreamPoolGetDataClient(getDataMetricTracker, getDataStreamPool);
+    HeartbeatSender heartbeatSender =
+        createStreamingEngineHeartbeatSender(
+            options, windmillServer, getDataStreamPool, configFetcher.getGlobalConfigHandle());
+    WorkCommitter workCommitter =
+        StreamingEngineWorkCommitter.builder()
+            .setCommitWorkStreamFactory(
+                WindmillStreamPool.create(
+                        numCommitThreads, COMMIT_STREAM_TIMEOUT, windmillServer::commitWorkStream)
+                    ::getCloseableStream)
+            .setCommitByteSemaphore(Commits.maxCommitByteSemaphore())
+            .setNumCommitSenders(numCommitThreads)
+            .setOnCommitComplete(this::onCompleteCommit)
+            .build();
+    GetWorkSender getWorkSender =
+        GetWorkSender.forStreamingEngine(
+            receiver -> windmillServer.getWorkStream(request, receiver));
+
+    this.getDataStatusProvider = getDataClient::printHtml;
+    this.currentActiveCommitBytesProvider = workCommitter::currentActiveCommitBytes;
+    this.channelzServlet =
+        createChannelzServlet(options, windmillServer::getWindmillServiceEndpoints);
+
+    return SingleSourceWorkerHarness.builder()
+        .setStreamingWorkScheduler(streamingWorkScheduler)
+        .setWorkCommitter(workCommitter)
+        .setGetDataClient(getDataClient)
+        .setComputationStateFetcher(this.computationStateCache::get)
+        .setWaitForResources(() -> memoryMonitor.waitForResources("GetWork"))
+        .setHeartbeatSender(heartbeatSender)
+        .setGetWorkSender(getWorkSender)
+        .build();
+  }
+
+  private void switchStreamingWorkerHarness(
+      ConnectivityType connectivityType,
+      long clientId,
+      DataflowWorkerHarnessOptions options,
+      GrpcWindmillStreamFactory windmillStreamFactory,
+      StreamingWorkScheduler streamingWorkScheduler,
+      ThrottlingGetDataMetricTracker getDataMetricTracker,
+      MemoryMonitor memoryMonitor,
+      ChannelCache channelCache,
+      GrpcDispatcherClient dispatcherClient,
+      WindmillServerStub windmillServer) {
+    if (connectivityType == ConnectivityType.CONNECTIVITY_TYPE_DIRECTPATH) {
+      if (!(this.streamingWorkerHarness.get() instanceof FanOutStreamingEngineWorkerHarness)) {
+        LOG.debug("Shutting down to SingleSourceWorkerHarness");
+        this.streamingWorkerHarness.get().shutdown();
+        FanOutStreamingEngineWorkerHarness fanoutStreamingWorkerHarness =
+            createFanOutStreamingEngineWorkerHarness(
+                clientId,
+                options,
+                windmillStreamFactory,
+                streamingWorkScheduler,
+                getDataMetricTracker,
+                memoryMonitor,
+                channelCache,
+                dispatcherClient);
+        this.streamingWorkerHarness.set(fanoutStreamingWorkerHarness);
+        this.getDataStatusProvider = getDataMetricTracker::printHtml;
```

Comment:
resetting these here won't update the references inside `statusPages`. Can we update the references in statusPages too?